OpenEdXMobile/src/main/java/org/edx/mobile/view/dialog/CourseModalDialogFragment.kt

```diff
@@ -105,175 +99,133 @@ class CourseModalDialogFragment : DialogFragment() {
-                    }
-                    HttpStatus.NOT_ACCEPTABLE -> {
-                        iapUtils.showPostUpgradeErrorDialog(
-                            context = this@CourseModalDialogFragment,
-                            errorResId = errorMsg.errorResId,
-                            errorCode = errorMsg.throwable.httpErrorCode,
-                            errorMessage = errorMsg.throwable.errorMessage,
-                            errorType = errorMsg.errorCode,
-                            retryListener = { _, _ ->
-                                iapViewModel.upgradeMode =
-                                    InAppPurchasesViewModel.UpgradeMode.SILENT
-                                iapViewModel.showFullScreenLoader(true)
-                                dismiss()
-                            },
-                            cancelListener = null
-                        )
-                    }
-                    else -> iapUtils.showUpgradeErrorDialog(
-                        context = this@CourseModalDialogFragment,
-                        errorResId = errorMsg.errorResId,
-                        errorCode = errorMsg.throwable.httpErrorCode,
-                        errorMessage = errorMsg.throwable.errorMessage,
-                        errorType = errorMsg.errorCode
-                    )
-                }
+                handleIAPException(errorMsg)
             } else {
-                iapUtils.showUpgradeErrorDialog(
+                iapDialog.showUpgradeErrorDialog(
                     context = this@CourseModalDialogFragment,
                     errorResId = errorMsg.errorResId,
                     errorType = errorMsg.errorCode
                 )
             }
             iapViewModel.errorMessageShown()
         })
-    }
 
-    private fun enableUpgradeButton(enable: Boolean) {
-        binding.layoutUpgradeBtn.btnUpgrade.setVisibility(enable)
-        binding.layoutUpgradeBtn.loadingIndicator.setVisibility(!enable)
+        iapViewModel.productPurchased.observe(viewLifecycleOwner, NonNullObserver {
+            lifecycleScope.launch {
+                iapViewModel.showFullScreenLoader(true)
+                dismiss()
+            }
+        })
     }
 
-    private fun purchaseProduct(productId: String) {
-        activity?.let { context ->
+    private fun setUpUpgradeButton(skuDetails: SkuDetails) {
+        binding.layoutUpgradeBtn.btnUpgrade.text =
+            ResourceUtil.getFormattedString(
+                resources,
+                R.string.label_upgrade_course_button,
+                AppConstants.PRICE,
+                skuDetails.price
+            ).toString()
+        // The app get the sku details instantly, so add some wait to perform
+        // animation at least one cycle.
+        binding.layoutUpgradeBtn.shimmerViewContainer.postDelayed({
+            binding.layoutUpgradeBtn.shimmerViewContainer.hideShimmer()
+            binding.layoutUpgradeBtn.btnUpgrade.isEnabled = true
```

Comment:
better to trigger the PriceLoad event from this point instead of when we get the price from the billing processor,
iapAnalytics.trackIAPEvent(Events.IAP_LOAD_PRICE_TIME)